File: Baekjoon/dongwan/A와 B 2.py
S = input()
T = input()

n = len(T)
def solution(x) :
    #T 길이에 도달했을 때, 일치하면 1 출력 후 종료 
    if len(x) == n :
        if x == T :
            print(1)
            exit()
        else :
            return 
    #A를 붙여보고 
    sol1 = x+'A'
    #들어갈 수 있는지 확인 들어갈 가능성이 있으면 재귀 
    if sol1 in T or sol1[::-1] in T :
        solution(sol1)
    #B를 붙여보고 확인 
    sol2 = x+'B'
    if sol2 in T or sol2[::-1] in T :
        solution(sol2[::-1])


solution(S) 

print(0)

File: Baekjoon/jaehyung/A와B2.java
package Baekjoon.jaehyung;

/**
 * A와B 2 [골드 5] (성공)
 * https://www.acmicpc.net/problem/12919
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class A와B2 {

  private void dfs(String S, String T){
    System.out.println(T);
    if (T.length() < S.length()){
      return;
    }
    if (S.length() == T.length()){
      if (S.equals(T)){
        System.out.println(1);
        System.exit(0);
      }
    } else {
      if (T.charAt(0) == 'B'){
        dfs(S, removeB(T));
      }
      if (T.charAt(T.length()-1) == 'A'){
        dfs(S, removeA(T));
      }
    }
  }

  private String removeB(String t){
    StringBuilder sb = new StringBuilder(t.substring(1));
    sb.reverse();
    return sb.toString();
  }

  private String removeA(String t){
    return t.substring(0, t.length()-1);
  }

  public void solution(String S, String T) {
    if (S.length() > T.length()){
      System.out.println(0);
    } else if (S.length() == T.length()){
      if (S.equals(T)){
        System.out.println(1);
      } else {
        System.out.println(0);
      }
    } else {
      dfs(S, T);
      System.out.println(0);
    }
  }

  public static void main(String[] args) throws IOException {
    A와B2 TT = new A와B2();
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    String S = br.readLine();
    String T = br.readLine();

    TT.solution(S, T);
  }
}
